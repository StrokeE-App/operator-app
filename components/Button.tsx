'use client';
//A button component that recieves a title, click event and a color

interface ButtonProps {
	title: string;
	onClick: () => void;
	color: string;
	disabled?: boolean;
}

export default function Button({title, onClick, color, disabled = false}: ButtonProps) {
	const buttonColor = color === 'red' ? 'bg-customRed' : 'bg-customGreen';
	const textColor = color === 'red' ? 'text-customWhite' : 'text-customBlack';
	const disabledStyle = disabled ? 'opacity-50 cursor-not-allowed' : 'cursor-pointer';

	return (
		<button onClick={onClick} disabled={disabled} className={`${buttonColor} ${textColor} font-bold py-2 px-4 rounded ${disabledStyle}`}>
			{title}
		</button>
	);
}
