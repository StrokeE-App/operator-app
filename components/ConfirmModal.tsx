'use client';

import {useEffect, useState} from 'react';
import {Ambulance} from '@/types/ambulance';

interface ConfirmModalProps {
	isOpen: boolean;
	onClose: () => void;
	onConfirm: (selectedAmbulance?: string) => void;
	title: string;
	showDropdown?: boolean;
	ambulances?: Ambulance[];
}

export default function ConfirmModal({isOpen, onClose, onConfirm, title, showDropdown, ambulances}: ConfirmModalProps) {
	const [selectedAmbulance, setSelectedAmbulance] = useState('');

	// Handle the escape key to close the modal
	useEffect(() => {
		const handleEscape = (e: KeyboardEvent) => {
			if (e.key === 'Escape') onClose();
		};

		if (isOpen) {
			document.addEventListener('keydown', handleEscape);
			document.body.style.overflow = 'hidden';
		}

		return () => {
			document.removeEventListener('keydown', handleEscape);
			document.body.style.overflow = 'unset';
		};
	}, [isOpen, onClose]);

	if (!isOpen) return null;

	return (
		<div className="fixed inset-0 z-50 flex items-center justify-center">
			{/* Backdrop */}
			<div className="fixed inset-0 bg-black/50" onClick={onClose} aria-hidden="true" />

			{/* Modal */}
			<div className="relative bg-white rounded-lg w-[90%] max-w-sm p-6 shadow-lg" role="dialog" aria-modal="true">
				<div className="text-center space-y-6">
					<h2 className="text-lg font-medium">{title}</h2>

					{showDropdown && ambulances && (
						<select className="w-full p-2 mb-4 border rounded" value={selectedAmbulance} onChange={(e) => setSelectedAmbulance(e.target.value)}>
							<option value="">Seleccionar ambulancia</option>
							{ambulances.map((amb) => (
								<option key={amb.id} value={amb.id}>
									{amb.name}
								</option>
							))}
						</select>
					)}

					<div className="flex gap-4 justify-center">
						<button
							onClick={() => onConfirm(selectedAmbulance)}
							className="px-8 py-2 rounded-md bg-customGreen text-customBlack font-bold hover:bg-green-200 transition-colors disabled:bg-gray-50 "
							disabled={showDropdown && !selectedAmbulance}
						>
							Confirmar
						</button>
						<button onClick={onClose} className="px-8 py-2 rounded-md border border-red-500 text-red-500 hover:bg-red-50 transition-colors">
							Cancelar
						</button>
					</div>
				</div>
			</div>
		</div>
	);
}
