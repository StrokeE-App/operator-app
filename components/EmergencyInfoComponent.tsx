'use client';

import {EmergencyInfo} from '@/types';

type EmergencyInfoProps = {
	emergency: EmergencyInfo | null;
};

export default function EmergencyInfoComponent({emergency}: EmergencyInfoProps) {
	if (!emergency) {
		return (
			<>
				<div className="w-11/12 mx-auto p-6 ">
					<div className="text-center space-y-6">
						<div className="pb-4">
							<h1 className="text-2xl font-bold inline-block px-4 pb-1">Cargando...</h1>
						</div>
					</div>
				</div>
			</>
		);
	}

	const formatDate = (date: Date | string | null | undefined) => {
		if (!date) return 'Ahorita';

		try {
			// Convert string to Date if it's a string
			const dateObj = typeof date === 'string' ? new Date(date) : date;

			// Check if dateObj is a valid date
			if (isNaN(dateObj.getTime())) {
				throw new Error('Invalid date');
			}

			return new Intl.DateTimeFormat('es-CO', {
				year: 'numeric',
				month: 'short',
				day: 'numeric',
				hour: '2-digit',
				minute: '2-digit',
				second: '2-digit',
			}).format(dateObj);
		} catch (error) {
			console.error('Error formatting date:', error, 'Original value:', date);
			return 'Fecha desconocida';
		}
	};

	const readableStartDate = formatDate(emergency.startDate);
	const {firstName, lastName, phoneNumber, age, weight, height} = emergency.patient;
	return (
		<div className="w-11/12 mx-auto p-6 ">
			<div className="text-center space-y-6">
				<div className="pb-4">
					<h1 className="text-2xl font-bold inline-block px-4 pb-1">
						{firstName} {lastName}
					</h1>
				</div>

				<div className="grid grid-cols-2 gap-x-6 gap-y-6">
					<div>
						<h2 className="text-custom-black font-bold text-sm">Teléfono</h2>
						<p className="text-custom-black font-medium">{phoneNumber}</p>
					</div>
					<div>
						<h2 className="text-custom-black text-sm font-bold">Edad</h2>
						<p className="text-custom-black font-medium">{age} años</p>
					</div>
					<div>
						<h2 className="text-custom-black text-sm font-bold">Peso</h2>
						<p className="text-custom-black font-medium">{weight} kg</p>
					</div>
					<div>
						<h2 className="text-custom-black text-sm font-bold">Estatura</h2>
						<p className="text-custom-black font-medium">{height} m</p>
					</div>
					<div>
						<h2 className="text-custom-black text-sm font-bold">Nivel De Stroke</h2>
						<p className="text-custom-black font-medium">{emergency.nihScale ? emergency.nihScale : '...'}</p>
					</div>
					<div>
						<h2 className="text-custom-black text-sm font-bold whitespace-wrap">Tiempo desde que inició la emergencia</h2>
						<p className="text-custom-black font-medium">{readableStartDate}</p>
					</div>
				</div>
			</div>
		</div>
	);
}
