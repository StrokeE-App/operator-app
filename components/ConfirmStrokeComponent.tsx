'use client';

import React, {useState} from 'react';

// Components
import Button from './Button';
import ConfirmModal from './ConfirmModal';

// Mocks
import {mockAmbulances} from '../mocks/ambulances';
import apiClient from '@/api/apiClient';
import toast from 'react-hot-toast';
import {useRouter} from 'next/navigation';
import {AxiosError} from 'axios';

export type ConfirmStrokeComponentProps = {
	emergencyId: string;
};

export default function ConfirmStrokeComponent({emergencyId}: ConfirmStrokeComponentProps) {
	const [isModalOpen, setIsModalOpen] = useState(false);
	const [modalTitle, setModalTitle] = useState('');
	const [actionType, setActionType] = useState('');

	const router = useRouter();

	// Handle the confirm or discard action from the modal, and call the API
	const handleConfirm = async (selectedAmbulance?: string) => {
		const loadingToast = toast.loading('Cargando...');
		if (actionType === 'confirm') {
			try {
				await apiClient.post('/operator/assign-ambulance', {
					emergencyId,
					ambulanceId: selectedAmbulance,
				});

				toast.success('Emergencia confirmada exitosamente', {id: loadingToast});
				router.push('/dashboard');
			} catch (error) {
				if (error instanceof AxiosError) {
					toast.error(error.response?.data.message, {id: loadingToast});
				} else {
					toast.error('Error al confirmar la emergencia', {id: loadingToast});
				}
				console.error('Error confirming emergency:', error);
			}
		} else if (actionType === 'discard') {
			try {
				await apiClient.post('/operator/cancel-emergency', {
					emergencyId,
				});

				toast.success('Emergencia descartada exitosamente', {id: loadingToast});
				router.push('/dashboard');
			} catch (error) {
				if (error instanceof AxiosError) {
					toast.error(error.response?.data.message, {id: loadingToast});
				} else {
					toast.error('Error al descartar la emergencia', {id: loadingToast});
				}
				console.error('Error discarding emergency:', error);
			}
		}
		setIsModalOpen(false);
	};

	// Open the modal with the title and action type
	const openModal = (title: string, action: string) => {
		setModalTitle(title); //
		setActionType(action);
		setIsModalOpen(true);
	};

	return (
		<div className="w-10/12 max-w-md mx-auto flex flex-col space-y-4 mb-5">
			<Button title="Confirmar Stroke" onClick={() => openModal('Asigna una ambulancia a la emergencia', 'confirm')} color="red" />
			<Button title="Descartar Stroke" onClick={() => openModal('¿Estás seguro que quieres descartar el stroke?', 'discard')} color="green" />
			<ConfirmModal
				isOpen={isModalOpen}
				onClose={() => setIsModalOpen(false)}
				onConfirm={handleConfirm}
				title={modalTitle}
				showDropdown={actionType === 'confirm'}
				ambulances={mockAmbulances}
			/>
		</div>
	);
}
