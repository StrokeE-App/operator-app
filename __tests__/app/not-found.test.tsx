import {render, screen} from '@testing-library/react';
import Custom404 from '@/app/not-found';
import {StrokeeLogo} from '@/components/StrokeeLogo';
import Link from 'next/link';

// Mock the components
jest.mock('@/components/StrokeeLogo', () => {
	return function MockStrokeeLogo() {
		return <div data-testid="strokee-logo">Strokee Logo</div>;
	};
});

jest.mock('next/link', () => {
	return function MockLink({children, href}: {children: React.ReactNode; href: string}) {
		return (
			<div data-testid="login-link" data-href={href}>
				{children}
			</div>
		);
	};
});

describe('Custom404', () => {
	it('should render all required elements', () => {
		render(<Custom404 />);

		expect(screen.getByTestId('strokee-logo')).toBeInTheDocument();
		expect(screen.getByText('Oops')).toBeInTheDocument();
		expect(screen.getByText('404 - No fue posible encontrar esta pagina :(')).toBeInTheDocument();
		expect(screen.getByText('Regresa al Login para continuar navegando:')).toBeInTheDocument();
		expect(screen.getByTestId('login-link')).toBeInTheDocument();
	});

	it('should have correct styling classes', () => {
		const {container} = render(<Custom404 />);

		// Check main container
		expect(container.querySelector('div')).toHaveClass('flex', 'flex-col', 'items-center', 'justify-center', 'h-screen');

		// Check heading styles
		expect(screen.getByText('Oops')).toHaveClass('text-customRed', 'text-4xl', 'font-bold');
		expect(screen.getByText('404 - No fue posible encontrar esta pagina :(')).toHaveClass('text-customRed', 'text-2xl', 'font-bold');
		expect(screen.getByText('Regresa al Login para continuar navegando:')).toHaveClass('text-customRed');
	});

	it('should have correct link to login page', () => {
		render(<Custom404 />);

		const loginLink = screen.getByTestId('login-link');
		expect(loginLink).toHaveAttribute('data-href', '/login');
		expect(loginLink).toHaveTextContent('Login');
	});

	it('should maintain consistent layout', () => {
		const {container} = render(<Custom404 />);

		// Check if elements are in the correct order
		const elements = container.querySelectorAll('div > *');
		expect(elements[0]).toHaveAttribute('data-testid', 'strokee-logo');
		expect(elements[1]).toBeEmptyDOMElement(); // <br />
		expect(elements[2]).toHaveTextContent('Oops');
		expect(elements[3]).toHaveTextContent('404 - No fue posible encontrar esta pagina :(');
		expect(elements[4]).toContainElement(screen.getByTestId('login-link'));
	});
});
