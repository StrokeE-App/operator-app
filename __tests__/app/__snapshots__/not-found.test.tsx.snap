// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom404 matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center h-screen"
  >
    <div
      data-testid="strokee-logo"
    >
      Strokee Logo
    </div>
    <br />
    <h1
      class="text-customRed text-4xl font-bold"
    >
      Oops
    </h1>
    <h2
      class="text-customRed text-2xl font-bold"
    >
      404 - No fue posible encontrar esta pagina :(
    </h2>
    <p
      class="text-customRed"
    >
      Regresa al Login para continuar navegando: 
      <a
        class="underline decoration-1"
        href="/login"
      >
        Login
      </a>
    </p>
  </div>
</div>
`;
